### ISSUE TITLE
2026 Q2 Roadmap

### ISSUE THREAD

**Issue by yueming-yuan** [radixark/miles#797]
### Training & Inference Alignment 
@maocheng23 
- [ ] Support GLM5/Kimi-K2.5/Qwen3.5
- [ ] Performance check for different models
- [ ] Prefill as draft model
- [ ] Refactor for better implementation later, add UTs and e2e tests

### Fully Async 
@maocheng23 
- [x] Support in_place/retract fully async run
- [x] Add staleness control on fully async
- [ ] More metrics to monitor fullt async performance

### On Policy Distillation
@maocheng23 
- [ ] Better OPD design on Miles
- [ ] Add OPD CI

### Token-In-Token-Out (TITO) 
@guapisolo 
- [ ] Support multi sample collection in a simple session.
- More details here #712

### New Model Support
@Zhichenzzz @guapisolo 
- [x] Nemotron 3 Super @Zhichenzzz
- [ ] gemma-4 @Zhichenzzz 
- [x] Kimi K2.5
  - [x] Original support from @GeLee-Q
  - [ ] Support mxfp8 training recipe on Blackwell
- [x] GLM 5 @guapisolo
- [x] Qwen 3.5 Series with real CP. @Zhichenzzz 
  - [ ] other model sizes: 0.8B, 2B, 122B-A10B, 397B-A17B
- [ ] MiMo-V2-Flash
- [ ] gpt-oss-120b
- [ ] Llama 4
- [ ] MiniMax-M2
- [ ] VL series
  - [ ] Qwen3.5-VL, Qwen3-VL 

### Agentic Training Examples & Infra
@Shi-Dong @BHZ-BER 
- [ ] Make sure that agentic training with GLM-4.7-Full or GLM-5 runs stably
- [ ] Harbor framework optimization for agentic training.  
- [ ] Nemo Gym integration and example runs
- [ ] Terminal agent and example runs (SETA, Endless Terminals)
- [ ] Prime Intellect verifiers

### Low-Precision training 
@yueming-yuan @zianglih
- [ ] #615 

### VLM Enhancement 
@yueming-yuan 
- [ ] Kimi K2.5 large-scale test
- [ ] VLM training experience enhancement

### LoRA training 
@yushengsu-thu 
##### P0
- [x] Disaggregated RL lora training mode
- [x] Weight sync optimization 
- [ ] Support more large-scale models: deepseek,  Kimi, Qwen
- [x] Optimize the LoRA inference kernel to achieve at least 50% of the base model inference speed 
- [ ] Quantization Lora (fp8) 
##### P1
- [ ] Qwen3.5 moe series support (training recipe) 
- [ ] Multi-lora RL training: multi-request, 1 trainer, multi inferences 
- [ ] Train-inference dismatch: integrate sgl `fused_moe_lora` kernels to Megatron or megatron-bridge 
##### P2 (academic)
- [ ] Optimize training trajectory: scheduler, optimizer (issue: slow convergence, poor acc, etc..)

### Miles-Diffusion 
@guapisolo 
- [ ] TODO

### Documentation
@Zhichenzzz @fzyzcjy @zhaochenyang20 
- [ ] TODO

### Experience Improvement
- [ ] Training script refactor, argument unify

**Comment by BHZ-BER** [comment]
Shall we also include some refactor plan here (if any)?

**Comment by zianglih** [comment]
Blackwell MXFP8 & NVFP4 roadmap: https://github.com/radixark/miles/issues/615

**Comment by Priyadlfw** [comment]
@kaixih want to include the [GB300 enablement](https://github.com/radixark/miles/pull/808#pullrequestreview-4011634586) here

**Comment by slopinator9000** [comment]
Gemma-4 is checked off as completed here, but the PR for it is still open https://github.com/radixark/miles/pull/1232.

Is it safe to check out the Gemma-4 PR to train Gemma-4 26B-4B, or is that still a work in progress?

**Comment by Zhichenzzz** [comment]
> Gemma-4 is checked off as completed here, but the PR for it is still open [#1232](https://github.com/radixark/miles/pull/1232).
> 
> Is it safe to check out the Gemma-4 PR to train Gemma-4 26B-4B, or is that still a work in progress?

Hi @slopinator9000, we have already merged the Gemma-4 PR into miles main, feel free to use it~ Thanks for waiting!

**Comment by slopinator9000** [comment]
That's great to hear @Zhichenzzz ! The PR mentions "Single-node RL training support". Is there anything in the PR that would not allow for multi-node training of Gemma-4?

**Comment by chi030303** [comment]
Claiming the Qwen3.5 **0.8B and 2B** training recipes only (not 122B-A10B / 397B-A17B).

Plan: clone `scripts/models/qwen3.5-4B.py` plus matching `scripts/run_qwen3_dense.py` entries, fill HF shapes (layers / hidden / heads), and smoke a few GRPO steps on 4×A100 40GB.
